src/GafferScene/Set.cpp

```diff
@@ -192,75 +194,105 @@ IECore::ConstInternedStringVectorDataPtr Set::computeSetNames( const Gaffer::Con
 {
 	ConstInternedStringVectorDataPtr inNamesData = inPlug()->setNamesPlug()->getValue();
 
-	InternedString name = namePlug()->getValue();
-	if( !name.string().size() )
+	const std::string &names = namePlug()->getValue();
+	if( !names.size() )
 	{
 		return inNamesData;
 	}
 
-	const std::vector<InternedString> &inNames = inNamesData->readable();
-	if( std::find( inNames.begin(), inNames.end(), name ) != inNames.end() )
+	if( modePlug()->getValue() == Remove )
 	{
 		return inNamesData;
 	}
 
-	if( modePlug()->getValue() == Remove )
+	InternedStringVectorDataPtr result = NULL;
+
+	Tokenizer namesTokenizer( names, boost::char_separator<char>( " " ) );
+	const std::vector<InternedString> &inNames = inNamesData->readable();
+	std::vector<InternedString>::const_iterator inBegin = inNames.begin();
+	std::vector<InternedString>::const_iterator inEnd = inNames.end();
+	for( Tokenizer::iterator name = namesTokenizer.begin(), end = namesTokenizer.end(); name != end; name++ )
+	{
+		if( std::find( inBegin, inEnd, *name ) != inNames.end() )
+		{
+			continue;
+		}
+
+		if ( result == NULL )
+		{
+			// we are modifying the list, make a copy now
+			result = inNamesData->copy();
+		}
+
+		result->writable().push_back( *name );
+	}
+
+	if ( result == NULL )
 	{
+		// output is the same as the input
 		return inNamesData;
 	}
 
-	InternedStringVectorDataPtr result = inNamesData->copy();
-	result->writable().push_back( name );
 	return result;
 }
 
 void Set::hashSet( const IECore::InternedString &setName, const Gaffer::Context *context, const ScenePlug *parent, IECore::MurmurHash &h ) const
 {
-	if( setName != namePlug()->getValue() )
+	Tokenizer namesTokenizer( namePlug()->getValue(), boost::char_separator<char>( " " ) );
+	for( Tokenizer::iterator name=namesTokenizer.begin(); name!=namesTokenizer.end(); name++ )
 	{
-		h = inPlug()->setPlug()->hash();
-		return;
+		if( setName == *name )
+		{
+			SceneProcessor::hashSet( setName, context, parent, h );
+			inPlug()->setPlug()->hash( h );
+			modePlug()->hash( h );
+			pathMatcherPlug()->hash( h );
+			return;
+		}
 	}
 
-	SceneProcessor::hashSet( setName, context, parent, h );
-	inPlug()->setPlug()->hash( h );
-	modePlug()->hash( h );
-	pathMatcherPlug()->hash( h );
+	// did not find the setName in any of the names, so simply return the hash from the input
+	h = inPlug()->setPlug()->hash();
 }
 
 GafferScene::ConstPathMatcherDataPtr Set::computeSet( const IECore::InternedString &setName, const Gaffer::Context *context, const ScenePlug *parent ) const
 {
-	if( InternedString( namePlug()->getValue() ) != setName )
-	{
-		return inPlug()->setPlug()->getValue();
-	}
-
-	ConstPathMatcherDataPtr pathMatcher = pathMatcherPlug()->getValue();
-	switch( modePlug()->getValue() )
+	Tokenizer namesTokenizer( namePlug()->getValue(), boost::char_separator<char>( " " ) );
```

Comment:
Again, let's avoid boost::tokenizer. Also, let's keep the separation between the early out case (no matching name) and the doing the work case, to avoid the unnecessary level of nesting. We can use `std::find()` to avoid the loop anyway, and it will all read much more easily.